### PR TITLE
Parsing support for jsDocComments.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
These are the same code changes as in https://github.com/Microsoft/TypeScript/pull/2646.  However, this only contains the changes necessary for parsing, and none of the binding/checking changes.

Those will be coming in later once we agree on an overall design for how we incorporate the type information from those elements.

In the meantime though, supporting parsing is simple and will enable nice features like classification/completion in js doc comments.

### CODE REVIEW

File: src/compiler/core.ts [hunk 1]
```diff
@@ -312,8 +312,11 @@ module ts {
 
         Debug.assert(start >= 0, "start must be non-negative, is " + start);
         Debug.assert(length >= 0, "length must be non-negative, is " + length);
-        Debug.assert(start <= file.text.length, `start must be within the bounds of the file. ${ start } > ${ file.text.length }`);
-        Debug.assert(end <= file.text.length, `end must be the bounds of the file. ${ end } > ${ file.text.length }`);
+
+        if (file) {
```
Comment (JsonFreeman) [inline src/compiler/core.ts]: Why would file be undefined?

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -468,6 +468,12 @@ module ts {
         return false;
     }
 
+    export function isClassLike(node: Node): boolean {
```
Comment (JsonFreeman) [inline src/compiler/utilities.ts]: Or just isClass. I'm getting a little tired of the "like" names

**Comment by mhegazy** [comment]
I do not see any tests for error cases.. are these going to be added later?

File: src/compiler/parser.ts [hunk 3]
```diff
@@ -2,6 +2,8 @@
 /// <reference path="utilities.ts"/>
 
 module ts {
+    export var throwOnJSDocErrors = false;
```
Comment (mhegazy) [inline src/compiler/parser.ts]: when is this set to true?

Comment (CyrusNajmabadi) [inline src/compiler/parser.ts]: This has been removed.  It was only for ascertaining what we could'nt handle on code in the wild.

**Comment by CyrusNajmabadi** [comment]
Also, there are about 32 tests for the error cases.

**Pull request merged**